File: 11-container-with-most-water/11-container-with-most-water.go
/*
Time - O(n)
Space - O(1)
Two Pointers
Greedy approach
*/
func maxArea(height []int) int {
    lo:=0
    hi:=len(height)-1
    maxArea:=0
    for lo<hi {
        area:=min(height[lo],height[hi])*(hi-lo)
        if area>maxArea {
            maxArea=area
        }
        if height[lo]>height[hi] {
            hi--
        }else {
            lo++
        }
    }
    return maxArea
}

func min(i,j int) int {
    if i<j {
        return i
    }
    return j
} 

File: 11-container-with-most-water/11-container-with-most-water.py
#Time - O(n)
#Space - O(1)

class Solution:
    def maxArea(self, height: List[int]) -> int:
        lo = 0
        hi = len(height)-1
        maxArea = 0
        while lo<hi:
            if height[lo]<height[hi]:
                area = height[lo]*(hi-lo)
                lo+=1
            else:
                area = height[hi]*(hi-lo)
                hi-=1
            if area>maxArea:
                maxArea=area
        return maxArea
                    
        